zxlive/commands.py

```diff
@@ -89,17 +89,42 @@ class ChangeNodeColor(BaseCommand):
     vty: VertexType.Type
 
     _old_vtys: Optional[list[VertexType]] = field(default=None, init=False)
+    _old_w_info: Optional[dict] = field(default=None, init=False)
 
     def undo(self) -> None:
         assert self._old_vtys is not None
         for v, old_vty in zip(self.vs, self._old_vtys):  # TODO: strict=True in Python 3.10
+            if vertex_is_w(old_vty):
+                v2 = self._old_w_info[v]['partner']
+                self.g.add_vertex_indexed(v2)
+                self.g.set_type(v2, self._old_w_info[v]['partner_type'])
+                self.g.set_row(v2, self._old_w_info[v]['partner_pos'][0])
+                self.g.set_qubit(v2, self._old_w_info[v]['partner_pos'][1])
+                self.g.add_edge(self.g.edge(v,v2), edgetype=EdgeType.W_IO)
+                for v3 in self._old_w_info[v]['neighbors']:
+                    self.g.add_edge(self.g.edge(v2,v3), edgetype=self.g.edge_type(self.g.edge(v,v3)))
+                    self.g.remove_edge(self.g.edge(v,v3))
             self.g.set_type(v, old_vty)
         self.update_graph_view()
 
     def redo(self) -> None:
+        if self._old_w_info is None:
+            self._old_w_info = {}
+        if vertex_is_w(self.vty):
+            return
         self._old_vtys = [self.g.type(v) for v in self.vs]
-        for v in self.vs:
-            self.g.set_type(v, self.vty)
+        for v1 in self.vs:
+            if vertex_is_w(self.g.type(v1)):
+                v2 = get_w_partner(self.g, v1)
+                v2_neighbors = [v for v in self.g.neighbors(v2) if v != v1]
+                for v3 in v2_neighbors:
+                    self.g.add_edge(self.g.edge(v1,v3), edgetype=self.g.edge_type(self.g.edge(v2,v3)))
+                self._old_w_info[v1] = {'partner': v2,
+                                        'partner_type': self.g.type(v2),
+                                        'partner_pos': (self.g.row(v2), self.g.qubit(v2)),
+                                        'neighbors': v2_neighbors}
+                self.g.remove_vertex(v2)
+            self.g.set_type(v1, self.vty)
```

Comment:
Now that this type change can also be a W node, we should rename the command from `ChangeNodeColor` to something like `ChangeNodeType`